### PR TITLE
[MISC] Extend insert_iterator

### PR DESCRIPTION
**Pull request by eseiler**
For our `compute_sketches` benchmark, this is around 10-14 times faster.

We could also add an overload for the ibf, which then just calls `emplace`.
This could be used in `insert_into_ibf` and the IBF ctor, where we currently first put the hashes in a set or vector before inserting.

Numbers are iterations.
Clang apparently optimizes way better than GCC.
`Clang with inline` = remove `[[gnu::noinline]]` from `void invoke_without_inlining`

| Benchmark                      | GCC before | GCC | Clang before | Clang | Clang with inline |
|--------------------------------|------------|-----|--------------|-------|-------------------|
| manual_construct/64            |         32 |  32 |           44 |    54 |                41 |
| manual_construct/128           |         34 |  34 |           52 |    54 |                42 |
| manual_construct/256           |         34 |  34 |           54 |    54 |                44 |
| manual_construct/512           |         33 |  33 |           51 |    54 |                44 |
| manual_construct/1024          |         34 |  28 |           52 |    54 |                44 |
| config_construct/64            |          8 |  21 |           10 |    28 |                14 |
| config_construct/128           |          9 |  20 |           11 |    30 |                16 |
| config_construct/256           |          9 |  20 |           11 |    30 |                16 |
| config_construct/512           |          9 |  18 |           11 |    29 |                16 |
| config_construct/1024          |          9 |  16 |           12 |    26 |                15 |
| config_and_max_construct/64    |         13 |  27 |           17 |    39 |                23 |
| config_and_max_construct/128   |         13 |  28 |           17 |    39 |                23 |
| config_and_max_construct/256   |         13 |  28 |           18 |    40 |                23 |
| config_and_max_construct/512   |         11 |  28 |           17 |    40 |                24 |
| config_and_max_construct/1024  |         13 |  27 |           18 |    40 |                24 |

### CODE REVIEW

**Comment by seqan-actions** [comment]
Documentation preview available at https://docs.seqan.de/preview/seqan/hibf/237

File: include/hibf/misc/insert_iterator.hpp [hunk 1]
```diff
@@ -29,32 +31,51 @@ class insert_iterator
...
+        ptr{std::addressof(ibf)},
+        ibf_bin_index{ibf_bin_index},
+        type{data_type::ibf}
     {}
 
-    insert_iterator & operator=(uint64_t const value) noexcept
+    explicit constexpr insert_iterator(function_t & fun) : ptr{std::addressof(fun)}, type{data_type::function}
+    {}
+
+    [[gnu::always_inline, gnu::flatten]] constexpr insert_iterator & operator=(uint64_t const value) noexcept
```
Comment (smehringer) [inline include/hibf/misc/insert_iterator.hpp:61]: How can this function be constexpr? The `type` in the switch case is not?

**Review by smehringer — COMMENTED** [review]
ah sorry! one more thing: When moving to a fast layout or partitioned IBF I need to compute not only hyperloglog sketches but at the same time minHash sketches. Before I had the kmer set and then computed the respective sketches from that. Now I could use the function type I guess? Or I could still use the set.

(min hash sketches is a small heap that gets updated with every new kmer for the XXX smallest kmers in the set).

File: include/hibf/misc/insert_iterator.hpp (hunk 1 above)
Comment (eseiler) [reply include/hibf/misc/insert_iterator.hpp]: The `type` would be known in a constant expression, but what isn't allowed in constexpr is casting from `void *`.
Made it inline instead.

**Comment by eseiler** [comment]
> ah sorry! one more thing: When moving to a fast layout or partitioned IBF I need to compute not only hyperloglog sketches but at the same time minHash sketches. Before I had the kmer set and then computed the respective sketches from that. Now I could use the function type I guess? Or I could still use the set.
> 
> (min hash sketches is a small heap that gets updated with every new kmer for the XXX smallest kmers in the set).

Didn't really get faster when using a function.
The main thing is that you cannot get rid of `kmers`, because you still need it.

**Review by smehringer — APPROVED** [review]

**Pull request merged**